### PR TITLE
Translations of algebraic properties via morphisms and left-inverses

### PR DESCRIPTION
**Pull request by MatthewDaggitt**
This is a generic version of some of the proofs found in #783. I'm also planning on using them to greatly simplify some upcoming proofs in for Rationals.

The module parameters to `Algebra.Morphism.RawMagma/RawMonoid` would be greatly simplified by the existence of a `IsLeftInverse` record (see #793). I've found it impossible to incorporate the existing `_LeftInverseOf_` module.

### CODE REVIEW

**Comment by JacquesCarette** [comment]
Hmm, the vast majority of this PR can be (should be?) automatically generated. @alhassy here is a test case for you.

**Comment by MatthewDaggitt** [comment]
> Hmm, the vast majority of this PR can be (should be?) automatically generated.

That would be magic :smile:

**Comment by MatthewDaggitt** [comment]
Blocked by #793

File: src/Algebra/Morphism/RawMagma.agda [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+open RawMagma using (Carrier; _≈_)
+
+module Algebra.Morphism.RawMagma
+  {a b ℓ₁ ℓ₂} {From : RawMagma b ℓ₂} {To : RawMagma a ℓ₁}
+  {to : Carrier From → Carrier To}
+  (isRawMagmaMorphism : IsRawMagmaMorphism From To to)
+  {from : Carrier To → Carrier From}
+  (from-to : ∀ x → _≈_ From (from (to x)) x)
+  (from-cong : ∀ {x y} → _≈_ To x y → _≈_ From (from x) (from y))
+  where
```
Comment (rei1024) [inline src/Algebra/Morphism/RawMagma.agda]: I think `from`, `from-to` and `from-cong` can be replaced by `(to-injective : ∀ {x y} → _≈_ To (to x) (to y) → _≈_ From x y)`.

Comment (MatthewDaggitt) [reply src/Algebra/Morphism/RawMagma.agda]: Hmm I can't see how to prove the results without a left inverse. Maybe I'm being slow! How would the proof for `comm-homo` look?

Comment (rei1024) [reply src/Algebra/Morphism/RawMagma.agda]:
```
comm-homo : Commutative T._≈_ _⊕_ → Commutative F._≈_ _∙_
comm-homo comm x y = to-injective $ begin
  to (x ∙ y)  ≈⟨ ∙-homo x y ⟩
  to x ⊕ to y ≈⟨ comm (to x) (to y) ⟩
  to y ⊕ to x ≈˘⟨ ∙-homo y x ⟩
  to (y ∙ x)  ∎
  where open SetoidReasoning T-setoid
```

Comment (MatthewDaggitt) [reply src/Algebra/Morphism/RawMagma.agda]: Haha, oh yes. Obvious now you point it out! Thanks for the spot :smile:

**Comment by MatthewDaggitt** [comment]
Updated to use the new function hierarchy and included @WolframKahl's suggestion about having a special-purpose name for each syntax declaration.

**Pull request merged**